File: data/readable_format/fortran/file2112.f90
PROGRAM TestNMRV
  IMPLICIT NONE
  INTEGER :: N
  REAL :: H
  REAL, ALLOCATABLE :: Q(:), S(:), U(:)
  INTEGER :: i

  ! Test parameters
  N = 5
  H = 0.1
  ALLOCATE(Q(N), S(N), U(N))
  
  ! Initialize test data
  Q = (/1.0, 2.0, 3.0, 4.0, 5.0/)
  S = (/2.0, 3.0, 4.0, 5.0, 6.0/)
  U = (/0.0, 0.0, 0.0, 0.0, 0.0/)
  
  ! Call the subroutine
  CALL NMRV(N, H, Q, S, U)
  
  ! Output for verification
  PRINT *, 'Output U:'
  DO i = 1, N
     PRINT *, U(i)
  END DO

END PROGRAM TestNMRV

SUBROUTINE NMRV(N, H, Q, S, U)
  IMPLICIT NONE
  INTEGER, INTENT(IN) :: N
  REAL, INTENT(IN) :: H
  REAL, INTENT(IN) :: Q(N), S(N)
  REAL, INTENT(OUT) :: U(N)
  INTEGER :: I
  REAL :: G, C0, C1, C2, D, UTMP

  G = H*H/12.0

  DO I = 2, N-1
    C0 = 1.0 + G*((Q(I-1) - Q(I+1))/2.0 + Q(I))
    C1 = 2.0 - G*(Q(I+1) + Q(I-1) + 8.0*Q(I))
    C2 = 1.0 + G*((Q(I+1) - Q(I-1))/2.0 + Q(I))
    D  = G*(S(I+1) + S(I-1) + 10.0*S(I))
    UTMP = C1*U(I) - C0*U(I-1) + D
    U(I+1) = UTMP / C2
  END DO
END SUBROUTINE NMRV

File: data/readable_format/cpp/file2112.cpp
#include <iostream>
#include <vector>

// Function declaration is optional here since the definition is in the same file
void NMRV(int N, double H, const std::vector<double>& Q, const std::vector<double>& S, std::vector<double>& U) {
    double G = H * H / 12.0;
    
    for (int i = 1; i < N-1; ++i) {  // Adjusted for 0-based indexing
        double C0 = 1.0 + G * ((Q[i-1] - Q[i+1]) / 2.0 + Q[i]);
        double C1 = 2.0 - G * (Q[i+1] + Q[i-1] + 8.0 * Q[i]);
        double C2 = 1.0 + G * ((Q[i+1] - Q[i-1]) / 2.0 + Q[i]);
        double D = G * (S[i+1] + S[i-1] + 10.0 * S[i]);
        double UTMP = C1 * U[i] - C0 * U[i-1] + D;
        U[i+1] = UTMP / C2;  // Adjusted for 0-based indexing
    }
}

int main() {
    int N = 5;
    double H = 0.1;
    std::vector<double> Q = {1.0, 2.0, 3.0, 4.0, 5.0};
    std::vector<double> S = {2.0, 3.0, 4.0, 5.0, 6.0};
    std::vector<double> U(N, 0.0); // Initialize U with zeros

    NMRV(N, H, Q, S, U);

    // Output for verification
    std::cout << "Output U:" << std::endl;
    for (double u : U) {
        std::cout << u << std::endl;
    }

    return 0;
}